ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/PollingSubscription.java

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.ksql.rest.server.resources.streaming;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+
+import org.apache.kafka.connect.data.Schema;
+
+import java.util.concurrent.TimeUnit;
+
+public abstract class PollingSubscription<T> implements Flow.Subscription {
+
+  private static final int BACKOFF_DELAY_MS = 100;
+
+  private final Flow.Subscriber<T> subscriber;
+  private final ListeningScheduledExecutorService exec;
+  private final Schema schema;
+
+  private boolean needsSchema = true;
+  private volatile boolean done = false;
+  private Throwable exception = null;
+  private boolean draining = false;
+  private volatile ListenableFuture future;
+
+  public PollingSubscription(
+      ListeningScheduledExecutorService exec,
+      Flow.Subscriber<T> subscriber,
+      Schema schema
+  ) {
+    this.exec = exec;
+    this.subscriber = subscriber;
+    this.schema = schema;
+  }
+
+  @Override
+  public void cancel() {
+    if (future != null) {
+      future.cancel(false);
+    }
+    exec.submit(this::close);
+  }
+
+  @Override
+  public void request(long n) {
+    if (needsSchema) {
+      if (schema != null) {
+        subscriber.onSchema(schema);
+      }
+      needsSchema = false;
+    }
+    // check status since request can be reentrant
+    if (!draining) {
```

Comment:
nit: this would be better named something like `isClosing`. It seems like things are being drained continuously, until the time comes for the subscription to close.